src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs

```diff
@@ -128,10 +104,10 @@ private static Task ForAsync<T>(T fromInclusive, T toExclusive, int dop, TaskSch
                     // Continue to loop while there are more elements to be processed.
                     while (!state.Cancellation.IsCancellationRequested)
                     {
-                        // Get the next element from the enumerator. For some types, we can get the next element with just
+                        // Get the next element from the enumerator. For primitive types, we can get the next element with just
                         // interlocked operations, avoiding the need to take a lock.  For other types, we need to take a lock.
                         T element;
-                        if (Interlockable())
+                        if (typeof(T).IsPrimitive)
```

Comment:
This will change the behaviour for bools and floating point I think.

Reply:
T is constrained to `IBinaryInteger<T>`.